src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioSolutionCloseTracker.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.ProjectSystem.VS;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [Export(typeof(SolutionCloseTracker))]
+    [System.Composition.Shared]
+    internal class VisualStudioSolutionCloseTracker : SolutionCloseTracker, IVsSolutionEvents, IDisposable
+    {
+        private uint _cookie;
+        private readonly IVsSolution _solution;
+
+        [ImportingConstructor]
+        public VisualStudioSolutionCloseTracker(
+           [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if (serviceProvider.GetService(typeof(SVsSolution)) is IVsSolution solution)
+            {
+                _solution = solution;
+                _solution.AdviseSolutionEvents(this, out _cookie);
+            }
+        }
+
+        public void Dispose()
+        {
+            _solution.UnadviseSolutionEvents(_cookie);
+        }
+
+        public int OnAfterOpenProject(IVsHierarchy pHierarchy, int fAdded)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnQueryCloseProject(IVsHierarchy pHierarchy, int fRemoving, ref int pfCancel)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnBeforeCloseProject(IVsHierarchy pHierarchy, int fRemoved)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnAfterLoadProject(IVsHierarchy pStubHierarchy, IVsHierarchy pRealHierarchy)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnQueryUnloadProject(IVsHierarchy pRealHierarchy, ref int pfCancel)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnBeforeUnloadProject(IVsHierarchy pRealHierarchy, IVsHierarchy pStubHierarchy)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnAfterOpenSolution(object pUnkReserved, int fNewSolution)
+        {
+            this.IsClosing = false;
+            return VSConstants.S_OK;
+        }
+
+        public int OnQueryCloseSolution(object pUnkReserved, ref int pfCancel)
+        {
+            return HResult.NotImplemented;
+        }
+
+        public int OnBeforeCloseSolution(object pUnkReserved)
+        {
+            this.IsClosing = true;
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterCloseSolution(object pUnkReserved)
+        {
+            this.IsClosing = false;
```

Comment:
Solution close doesn't mean VS close. There is a `Close Solution` menu item on the File menu so technically the "time after solution close" is infinite. Having said that, we could probably remove this line because we'd setting it to false in OnSolutionOpen.

> Wondering if we're guaranteed to have resolved all of our eventing 

I would assume no guarantees but this PR _shouldn't_ introduce any new issues around that. If anything by doing less work this means we would have gotten through more events just by virtue of CPU time.

Having said all of that, by making me type out "Solution close doesn't mean VS close" you made me realise that there is a huge glaring enormous humongous giant super-duper bug in this PR that I need to fix, in which the shared components never have their "solution is closing" field un-set, so thank you.